File: src/writer/services/epigraph.py
"""Pure-text epigraph detection helpers."""

from __future__ import annotations

from dataclasses import dataclass
import re


MAX_SCAN_LINES = 8
MAX_QUOTE_LINES = 4
MAX_TOTAL_LINES = 5
MAX_QUOTE_CHARS = 240
MAX_ATTRIBUTION_CHARS = 120

_DASH_PREFIX_RE = re.compile(r"^\s*[—-]{1,2}\s*")
_LEADING_QUOTES_RE = re.compile(r"^[\"'“”‘’「」『』《》〈〉‹›«»]+")
_TRAILING_QUOTES_RE = re.compile(r"[\"'“”‘’「」『』《》〈〉‹›«»]+$")
_BOOK_RE = re.compile(r"《[^《》\n]{1,40}》")
_ATTR_LINE_RE = re.compile(r"^\s*(?:[—-]{1,2}\s*)?(?P<content>.+?)\s*$")
_SAME_LINE_SPLIT_RE = re.compile(
    r"^(?P<quote>.+?)(?:\s{2,}|[ \t]+[—-]{1,2}[ \t]+)(?P<attr>.+)$"
)
_SAME_LINE_DASH_SPLIT_RE = re.compile(
    r"^(?P<quote>.+?)(?P<sep>[—-]{1,2}\s*)(?P<attr>.+)$"
)
_SAME_LINE_PUNCT_SPLIT_RE = re.compile(
    r"^(?P<quote>.+[。！？!?…][\"'”’」』》〉»]*)\s+(?P<attr>.+)$"
)
_SECTION_HEADING_RE = re.compile(
    r"^\s*(?:chapter|section|part|prologue|epilogue|第[一二三四五六七八九十百千0-9]+[章节回部卷篇])\b",
    re.IGNORECASE,
)


@dataclass(frozen=True)
class Epigraph:
    quote_text: str
    source_text: str
    raw_text: str
    end_offset: int

    @property
    def quote(self) -> str:
        return self.quote_text

    @property
    def attribution(self) -> str:
        return self.source_text

    @property
    def line_count(self) -> int:
        return _count_lines(self.raw_text)

    @property
    def source_span(self) -> tuple[int, int]:
        return (0, self.line_count)


def detect_epigraph(text: str) -> Epigraph | None:
    """Detect an opening epigraph from the first few lines of plain text."""
    if not text or not text.strip():
        return None

    scanned_lines = 0
    current_offset = 0
    line_ends: list[int] = []
    line_texts: list[str] = []

    while scanned_lines < MAX_SCAN_LINES and current_offset < len(text):
        newline_idx = text.find("\n", current_offset)
        if newline_idx == -1:
            line_end = len(text)
            next_offset = len(text)
        else:
            line_end = newline_idx
            next_offset = newline_idx + 1
        raw_line = text[current_offset:line_end]
        if raw_line.endswith("\r"):
            raw_line = raw_line[:-1]
        line_texts.append(raw_line)
        line_ends.append(next_offset)
        scanned_lines += 1
        if newline_idx == -1:
            break
        current_offset = next_offset

    scan_limit = len(line_texts)
    for attr_idx in range(scan_limit):
        candidate = _build_candidate(text, line_texts, line_ends, attr_idx)
        if candidate is not None:
            return candidate
    return None


def strip_epigraph(text: str) -> str:
    """Remove the leading epigraph block, if present, preserving the body."""
    epigraph = detect_epigraph(text)
    if epigraph is None:
        return text
    return text[epigraph.end_offset :].lstrip()


def _build_candidate(
    text: str,
    lines: list[str],
    line_ends: list[int],
    attr_idx: int,
) -> Epigraph | None:
    attr_line = lines[attr_idx]
    attr = _parse_attribution(attr_line)
    if attr is not None:
        quote_lines = lines[:attr_idx]
        candidate = _make_epigraph(
            text=text,
            quote_lines=quote_lines,
            attribution=attr,
            block_start=0,
            block_end=line_ends[attr_idx],
        )
        if candidate is not None:
            return candidate

    same_line = _parse_same_line(lines[attr_idx])
    if same_line is None or attr_idx > 0:
        return None
    quote_text, same_attr = same_line
    return _make_epigraph(
        text=text,
        quote_lines=[quote_text],
        attribution=same_attr,
        block_start=0,
        block_end=line_ends[attr_idx],
    )


def _make_epigraph(
    *,
    text: str,
    quote_lines: list[str],
    attribution: str,
    block_start: int,
    block_end: int,
) -> Epigraph | None:
    trimmed_quote_lines = _trim_blank_edges(quote_lines)
    if not trimmed_quote_lines:
        return None
    raw_text = text[block_start:block_end]
    line_count = _count_lines(raw_text)
    if len(trimmed_quote_lines) > MAX_QUOTE_LINES:
        return None
    if line_count > MAX_TOTAL_LINES:
        return None

    quote = "\n".join(line.rstrip() for line in trimmed_quote_lines).strip()
    source_text = attribution.strip()
    if not quote or not source_text:
        return None
    if len(quote) > MAX_QUOTE_CHARS or len(source_text) > MAX_ATTRIBUTION_CHARS:
        return None
    if not _looks_like_attribution(source_text):
        return None
    if _looks_like_body_opening(quote):
        return None

    return Epigraph(
        quote_text=quote,
        source_text=source_text,
        raw_text=raw_text,
        end_offset=block_end,
    )


def _parse_same_line(line: str) -> tuple[str, str] | None:
    stripped = line.strip()
    if not stripped:
        return None
    match = _SAME_LINE_SPLIT_RE.match(stripped)
    if not match:
        match = _SAME_LINE_DASH_SPLIT_RE.match(stripped)
    if not match:
        match = _SAME_LINE_PUNCT_SPLIT_RE.match(stripped)
    if not match:
        return None
    quote = match.group("quote").strip()
    attr = match.group("attr").strip()
    if not quote or not attr:
        return None
    if not _looks_like_attribution(attr):
        return None
    return quote, _normalise_attribution(attr)


def _parse_attribution(line: str) -> str | None:
    stripped = line.strip()
    if not stripped:
        return None
    match = _ATTR_LINE_RE.match(stripped)
    if not match:
        return None
    content = match.group("content").strip()
    if not _looks_like_attribution(content):
        return None
    return _normalise_attribution(content)


def _looks_like_attribution(text: str) -> bool:
    if not text or "\n" in text:
        return False
    compact = text.strip()
    if len(compact) < 2 or len(compact) > MAX_ATTRIBUTION_CHARS:
        return False
    if compact.endswith(("。", "！", "？", "!", "?")):
        return False
    if _BOOK_RE.search(compact):
        return True
    parts = [part.strip() for part in compact.split("，") if part.strip()]
    if len(parts) != 2:
        return False
    return any(_BOOK_RE.search(part) for part in parts)


def _normalise_attribution(text: str) -> str:
    compact = _DASH_PREFIX_RE.sub("", text.strip())
    return re.sub(r"\s+", " ", compact).strip()


def _looks_like_body_opening(quote: str) -> bool:
    stripped = quote.strip()
    if not stripped:
        return True
    single_line = "\n" not in stripped
    compact = stripped.replace("\n", " ").strip()
    plain = _strip_wrapping_quotes(compact)
    lines = [_strip_wrapping_quotes(line.strip()) for line in stripped.splitlines() if line.strip()]
    if not plain:
        return True
    if _SECTION_HEADING_RE.match(plain):
        return True
    if single_line and len(plain) > 90 and "。" not in plain and "." not in plain:
        return True
    if (
        len(lines) >= 2
        and all(len(line) >= 12 for line in lines)
        and sum(line.endswith(("。", "！", "？", ".", "!", "?")) for line in lines) >= 2
    ):
        return True
    if plain.endswith(("：", ":")):
        return True
    if plain.startswith(
        (
            "我",
            "我们",
            "那天",
            "后来",
            "于是",
            "首先",
            "今天",
            "清晨",
            "凌晨",
            "夜里",
            "雨",
            "风",
            "门外",
        )
    ) and len(plain) > 20:
        return True
    if "，" not in plain and "。" not in plain and len(plain.split()) > 18:
        return True
    return False


def _trim_blank_edges(lines: list[str]) -> list[str]:
    start = 0
    end = len(lines)
    while start < end and not lines[start].strip():
        start += 1
    while end > start and not lines[end - 1].strip():
        end -= 1
    return lines[start:end]


def _strip_wrapping_quotes(text: str) -> str:
    stripped = _LEADING_QUOTES_RE.sub("", text)
    stripped = _TRAILING_QUOTES_RE.sub("", stripped)
    return stripped.strip()


def _count_lines(text: str) -> int:
    return len(text.splitlines()) if text else 0

File: tauri-mvp/frontend/src/features/articles/epigraph.ts
export interface EpigraphParts {
  quote: string
  attribution: string
  body: string
  raw: string
}

const MAX_SCAN_LINES = 8
const MAX_QUOTE_LINES = 4
const MAX_TOTAL_LINES = 5
const MAX_QUOTE_CHARS = 240
const MAX_ATTRIBUTION_CHARS = 120
const BOOK_RE = /《[^《》\n]{1,40}》/
const DASH_PREFIX_RE = /^\s*[—-]{1,2}\s*/
const ATTR_LINE_RE = /^\s*(?:[—-]{1,2}\s*)?(.+?)\s*$/
const SAME_LINE_DASH_RE = /^(.+?)([—-]{1,2}\s*)(.+)$/
const SAME_LINE_SPACE_DASH_RE = /^(.+?)(?:\s{2,}|[ \t]+[—-]{1,2}[ \t]+)(.+)$/
const SAME_LINE_PUNCT_RE = /^(.+[。！？!?…]["'”’」』》〉»]*)\s+(.+)$/

export function detectEpigraph(text: string): EpigraphParts | null {
  if (!text.trim()) return null

  const lines: string[] = []
  const lineEnds: number[] = []
  let offset = 0
  while (lines.length < MAX_SCAN_LINES && offset < text.length) {
    const newline = text.indexOf('\n', offset)
    const lineEnd = newline === -1 ? text.length : newline
    const nextOffset = newline === -1 ? text.length : newline + 1
    let rawLine = text.slice(offset, lineEnd)
    if (rawLine.endsWith('\r')) rawLine = rawLine.slice(0, -1)
    lines.push(rawLine)
    lineEnds.push(nextOffset)
    if (newline === -1) break
    offset = nextOffset
  }

  for (let attrIndex = 0; attrIndex < lines.length; attrIndex += 1) {
    const attribution = parseAttribution(lines[attrIndex])
    if (attribution) {
      const candidate = makeEpigraph(text, lines.slice(0, attrIndex), attribution, lineEnds[attrIndex])
      if (candidate) return candidate
    }

    if (attrIndex === 0) {
      const sameLine = parseSameLine(lines[attrIndex])
      if (sameLine) {
        const candidate = makeEpigraph(text, [sameLine.quote], sameLine.attribution, lineEnds[attrIndex])
        if (candidate) return candidate
      }
    }
  }

  return null
}

export function composeArticleBody(parts: EpigraphParts | null, body: string): string {
  if (!parts || !parts.quote.trim() || !parts.attribution.trim()) {
    return body
  }
  return `${parts.quote.trim()}\n——${parts.attribution.trim()}\n\n${body.trimStart()}`
}

function makeEpigraph(text: string, quoteLines: string[], attribution: string, endOffset: number): EpigraphParts | null {
  const trimmedLines = trimBlankEdges(quoteLines)
  if (!trimmedLines.length || trimmedLines.length > MAX_QUOTE_LINES) return null

  const raw = text.slice(0, endOffset)
  const lineCount = countLines(raw)
  if (lineCount > MAX_TOTAL_LINES) return null

  const quote = trimmedLines.map((line) => line.trimEnd()).join('\n').trim()
  const normalizedAttribution = normalizeAttribution(attribution)
  if (!quote || !normalizedAttribution) return null
  if (quote.length > MAX_QUOTE_CHARS || normalizedAttribution.length > MAX_ATTRIBUTION_CHARS) return null
  if (!looksLikeAttribution(normalizedAttribution)) return null
  if (looksLikeBodyOpening(quote)) return null

  return {
    quote,
    attribution: normalizedAttribution,
    raw,
    body: text.slice(endOffset).trimStart(),
  }
}

function parseSameLine(line: string): { quote: string; attribution: string } | null {
  const stripped = line.trim()
  if (!stripped) return null
  const match = stripped.match(SAME_LINE_SPACE_DASH_RE)
    ?? stripped.match(SAME_LINE_DASH_RE)
    ?? stripped.match(SAME_LINE_PUNCT_RE)
  if (!match) return null
  const quote = match[1]?.trim()
  const attribution = match[match.length - 1]?.trim()
  if (!quote || !attribution || !looksLikeAttribution(attribution)) return null
  return { quote, attribution: normalizeAttribution(attribution) }
}

function parseAttribution(line: string): string | null {
  const stripped = line.trim()
  if (!stripped) return null
  const match = stripped.match(ATTR_LINE_RE)
  const content = match?.[1]?.trim()
  if (!content || !looksLikeAttribution(content)) return null
  return normalizeAttribution(content)
}

function looksLikeAttribution(text: string): boolean {
  const compact = text.trim()
  if (compact.length < 2 || compact.length > MAX_ATTRIBUTION_CHARS) return false
  if (/[。！？!?]$/.test(compact)) return false
  if (BOOK_RE.test(compact)) return true
  const parts = compact.split('，').map((part) => part.trim()).filter(Boolean)
  return parts.length === 2 && parts.some((part) => BOOK_RE.test(part))
}

function normalizeAttribution(text: string): string {
  return text.trim().replace(DASH_PREFIX_RE, '').replace(/\s+/g, ' ').trim()
}

function looksLikeBodyOpening(quote: string): boolean {
  const plain = quote.replace(/^[“”"'‘’「」『』《》〈〉]+/, '').replace(/[“”"'‘’「」『』《》〈〉]+$/, '').trim()
  if (!plain) return true
  if (/^\s*(chapter|section|part|prologue|epilogue|第[一二三四五六七八九十百千0-9]+[章节回部卷篇])/i.test(plain)) return true
  if (!plain.includes('\n') && plain.length > 90 && !/[。.]/.test(plain)) return true
  if (/[：:]$/.test(plain)) return true
  if (/^(我|我们|那天|后来|于是|首先|今天|清晨|凌晨|夜里|雨|风|门外)/.test(plain) && plain.length > 20) return true
  return false
}

function trimBlankEdges(lines: string[]): string[] {
  let start = 0
  let end = lines.length
  while (start < end && !lines[start].trim()) start += 1
  while (end > start && !lines[end - 1].trim()) end -= 1
  return lines.slice(start, end)
}

function countLines(text: string): number {
  if (!text) return 0
  const lines = text.split(/\r\n|\r|\n/)
  if (text.endsWith('\n') || text.endsWith('\r')) {
    return Math.max(0, lines.length - 1)
  }
  return lines.length
}
